source/common/conn_pool/conn_pool_base.cc

```diff
@@ -12,9 +12,17 @@ namespace ConnectionPool {
 ConnPoolImplBase::ConnPoolImplBase(
     Upstream::HostConstSharedPtr host, Upstream::ResourcePriority priority,
     Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
-    const Network::TransportSocketOptionsSharedPtr& transport_socket_options)
+    const Network::TransportSocketOptionsSharedPtr& transport_socket_options,
+    std::chrono::milliseconds pool_idle_timeout)
     : host_(host), priority_(priority), dispatcher_(dispatcher), socket_options_(options),
-      transport_socket_options_(transport_socket_options) {}
+      transport_socket_options_(transport_socket_options), idle_timeout_(pool_idle_timeout),
+      idle_timer_(dispatcher.createTimer([this]() {
```

Comment:
I have to look again every time I come back to this, but did you double check that conn pools are always deleted inline and not in deferred delete context? I think this is true and we can delete/disable the idle timer in the destructor but I want to make sure we don't have a potential issue where the timer might fire after the pool has been deferred deleted.

Reply:
It looks like the HTTP conn pools are never deferred deleted, but unfortunately the TCP conn pools are -- see [ClusterManagerImpl::ThreadLocalClusterManagerImpl::drainTcpConnPools](https://github.com/envoyproxy/envoy/blob/c8de78846a9c612cfff723b8edafec4c4189ada5/source/common/upstream/cluster_manager_impl.cc#L1085).

I can add a `cancelPoolIdleTimer()` method to `ConnectionPool::Instance` and add a call to it in `drainTcpConnPools`, unless there's a better pattern to follow.

Reply:
I would need to spend more time with the code to give you a better assessment here. It would be nice to have this be self contained somehow and not add more interface methods, but maybe that is not possible. cc @alyssawilk for additional thoughts as she has been in this code more recently.

Reply:
I think (please check) that drainConnectionsImpl is always called before pool destruction so that might be a good and safe place to unregister.

Reply:
I put the timer cancellation into `closeIdleConnectionsForDrainingPool` since that is called from all drain paths.